api/censuses.go

```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
 	qtx := capi.sqlc.WithTx(tx)
 	currentCensus, err := qtx.CensusByID(internalCtx, int64(censusID))
 	if err != nil {
+		log.Errorw(err, "error getting census from database")
 		if errors.Is(err, sql.ErrNoRows) {
 			return ErrNotFoundCensus
 		}
 		return ErrCantGetCensus
 	}
 	chainID, err := qtx.ChainID(internalCtx)
 	if err != nil {
+		log.Errorw(err, "error getting chainID")
 		return ErrCantGetCensus
 	}
+	censusSize := int32(0)
+	if currentCensus.Size.Valid {
+		censusSize = currentCensus.Size.Int32
+	}
+	censusWeight := []byte{}
+	if currentCensus.Weight.Valid {
+		censusWeight = []byte(currentCensus.Weight.String)
+	}
 	res, err := json.Marshal(GetCensusResponse{
 		CensusID:   uint64(censusID),
 		StrategyID: uint64(currentCensus.StrategyID),
 		MerkleRoot: common.Bytes2Hex(currentCensus.MerkleRoot),
 		URI:        "ipfs://" + currentCensus.Uri.String,
-		Size:       int32(currentCensus.Size),
-		Weight:     new(big.Int).SetBytes(currentCensus.Weight).String(),
+		Size:       int32(censusSize),
+		Weight:     new(big.Int).SetBytes(censusWeight).String(),
 		ChainID:    uint64(chainID),
+		Anonymous:  currentCensus.CensusType == int64(census.AnonymousCensusType),
 	})
 	if err != nil {
+		log.Errorw(err, "error encoding census")
 		return ErrEncodeCensus
 	}
 	return ctx.Send(res, api.HTTPstatusOK)
 }
 
-// createAndPublishCensus handler creates a census tree based on the token
-// holders of the tokens that are included in the given strategy. It recovers
-// all the required information from the database, and then creates and publish
-// the census merkle tree on IPFS. Then saves the resulting information of the
-// census tree in the database and returns its ID.
-//
-// TODO: This handler is costly, specially for big censuses. It should be refactored to be a background task.
-func (capi *census3API) createAndPublishCensus(msg *api.APIdata, ctx *httprouter.HTTPContext) error {
+// launchCensusCreation handler parses the census creation request, enqueues it
+// and starts the creation process, then returns the queue identifier of that
+// process to support tracking it. When the process ends updates the queue item
+// with the resulting status or error into the queue.
+func (capi *census3API) launchCensusCreation(msg *api.APIdata, ctx *httprouter.HTTPContext) error {
 	// decode request
 	req := &CreateCensusResquest{}
 	if err := json.Unmarshal(msg.Data, req); err != nil {
 		return ErrMalformedStrategyID
 	}
-	// get tokens associated to the strategy
-	internalCtx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
-	defer cancel()
+	// create and publish census merkle tree in background
+	queueID := capi.queue.Enqueue()
+	go func(req *CreateCensusResquest) {
```

Comment:
You don't need to pass req as parameter. The goroutine will inherit the vars from the caller routine.